pkg/compute/bidder.go

```diff
@@ -5,69 +5,104 @@ import (
+	SemanticStrategy bidstrategy.SemanticBidStrategy
+	ResourceStrategy bidstrategy.ResourceBidStrategy
+	Store            store.ExecutionStore
+	Callback         Callback
+	GetApproveURL    func() *url.URL
 }
 
 type Bidder struct {
 	nodeID        string
-	strategy      bidstrategy.BidStrategy
 	store         store.ExecutionStore
 	callback      Callback
 	getApproveURL func() *url.URL
+
+	semanticStrategy bidstrategy.SemanticBidStrategy
+	resourceStrategy bidstrategy.ResourceBidStrategy
 }
 
 func NewBidder(params BidderParams) Bidder {
 	return Bidder{
-		nodeID:        params.NodeID,
-		strategy:      params.Strategy,
-		store:         params.Store,
-		getApproveURL: params.GetApproveURL,
-		callback:      params.Callback,
+		nodeID:           params.NodeID,
+		store:            params.Store,
+		getApproveURL:    params.GetApproveURL,
+		callback:         params.Callback,
+		semanticStrategy: params.SemanticStrategy,
+		resourceStrategy: params.ResourceStrategy,
 	}
 }
 
-func (b Bidder) RunBidding(ctx context.Context, execution store.Execution) {
-	// ask the bidding strategy if we should bid on this job
-	bidStrategyRequest := bidstrategy.BidStrategyRequest{
-		NodeID:   b.nodeID,
-		Job:      execution.Job,
-		Callback: b.getApproveURL(),
-	}
+func (b Bidder) RunBidding(ctx context.Context, request AskForBidRequest, usageCalc capacity.UsageCalculator) {
+	go func() {
+		// ask the bidding strategy if we should bid on this job
+		bidStrategyRequest := bidstrategy.BidStrategyRequest{
+			NodeID:   b.nodeID,
+			Job:      request.Job,
+			Callback: b.getApproveURL(),
+		}
 
-	response, err := b.doBidding(ctx, bidStrategyRequest, execution.ResourceUsage)
-	if err != nil {
-		err = multierr.Append(err, b.store.UpdateExecutionState(ctx, store.UpdateExecutionStateRequest{
-			ExecutionID: execution.ID,
-			NewState:    store.ExecutionStateFailed,
-			Comment:     err.Error(),
-		}))
+		response, resourceUsage, err := b.doBidding(ctx, bidStrategyRequest, usageCalc)
+		if err != nil {
+			b.callback.OnComputeFailure(ctx, ComputeError{
+				RoutingMetadata: RoutingMetadata{
+					// TODO double check these.
+					SourcePeerID: b.nodeID,
+					TargetPeerID: request.SourcePeerID,
+				},
+				ExecutionMetadata: ExecutionMetadata{
+					ExecutionID: request.ExecutionID,
+					JobID:       request.JobID,
+				},
+				Err: err.Error(),
+			})
+
+			log.Ctx(ctx).Error().Err(err).Msg("Error running bid strategy")
+			return
+		}
 
-		b.callback.OnComputeFailure(ctx, ComputeError{
+		result := BidResult{
 			RoutingMetadata: RoutingMetadata{
 				SourcePeerID: b.nodeID,
-				TargetPeerID: execution.RequesterNodeID,
+				TargetPeerID: request.SourcePeerID,
 			},
-			ExecutionMetadata: NewExecutionMetadata(execution),
-			Err:               err.Error(),
-		})
+			ExecutionMetadata: ExecutionMetadata{
+				ExecutionID: request.ExecutionID,
+				JobID:       request.JobID,
+			},
+			Accepted: response.ShouldBid,
+			Reason:   response.Reason,
+		}
```

Comment:
We're now returning bid results from two places – one from here and one from ReturnBidResult, and they have subtly different content now. Is this a problem?

Reply:
It shouldn't be. The only place `ReturnBidResult` gets called is in https://github.com/bacalhau-project/bacalhau/blob/e866a9156c1384c29f13cee84cc4421a42eb345f/pkg/compute/publicapi/endpoint_approve.go#L22-L45

And we can be certain the `execution` exists in the store at that point if the bid was approved or is waiting. Perhaps we could name this method differently, or ideally remove it from the `Bidder` since the method has little to do with bidding in the first place.

Reply:
On second thought, this is a problem for our testing via mocked callbacks since it is unexpected to receive multiple `OnBidComplete` callbacks. Im going to give this a second look.

Reply:
Alright [sorted the testing](https://github.com/bacalhau-project/bacalhau/pull/2395/files#diff-e40c658cdd9c7b652d7597a1e363a9606dd596daef8e54d3b6491eef166837c3R61-R67). Tests calling `AskForBid` must wait for the execution to be created if the bid is accepted or waited on as I have done in the linked test package.